datafusion/core/src/physical_plan/file_format/row_filter.rs

```diff
@@ -0,0 +1,367 @@
+            projected_columns: false,
+        }
+    }
+
+    pub fn build(
+        mut self,
+        metadata: &ParquetMetaData,
+    ) -> Result<Option<FilterCandidate>> {
+        let expr = self.expr.clone();
+        let expr = expr.rewrite(&mut self)?;
+
+        if self.non_primitive_columns || self.projected_columns {
+            Ok(None)
+        } else {
+            let required_bytes =
+                size_of_columns(&self.required_columns, self.file_schema, metadata)?;
+            let can_use_index =
+                columns_sorted(&self.required_columns, self.file_schema, metadata)?;
+
+            Ok(Some(FilterCandidate {
+                expr,
+                required_bytes,
+                can_use_index,
+                projection: self.required_column_indices,
+            }))
+        }
+    }
+}
+
+impl<'a> ExprRewriter for FilterCandidateBuilder<'a> {
+    fn pre_visit(&mut self, expr: &Expr) -> Result<RewriteRecursion> {
+        if let Expr::Column(column) = expr {
+            if let Ok(idx) = self.file_schema.index_of(&column.name) {
+                self.required_columns.insert(column.clone());
+                self.required_column_indices.push(idx);
+
+                if !is_primitive_field(self.file_schema.field(idx)) {
+                    self.non_primitive_columns = true;
+                }
+            } else if self.table_schema.index_of(&column.name).is_err() {
+                // If the column does not exist in the (un-projected) table schema then
+                // it must be a projected column.
+                self.projected_columns = true;
+            }
+        }
+        Ok(RewriteRecursion::Continue)
+    }
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::Column(Column { name, .. }) = &expr {
+            if self.file_schema.field_with_name(name).is_err() {
+                return Ok(Expr::Literal(ScalarValue::Null));
+            }
+        }
+
+        Ok(expr)
+    }
+}
+
+/// Calculate the total compressed size of all `Column's required for
+/// predicate `Expr`. This should represent the total amount of file IO
+/// required to evaluate the predicate.
+fn size_of_columns(
+    columns: &HashSet<Column>,
+    schema: &Schema,
+    metadata: &ParquetMetaData,
+) -> Result<usize> {
+    let mut total_size = 0;
+    let row_groups = metadata.row_groups();
+    for Column { name, .. } in columns.iter() {
+        let idx = schema.index_of(name)?;
+
+        for rg in row_groups.iter() {
+            total_size += rg.column(idx).compressed_size() as usize;
+        }
+    }
+
+    Ok(total_size)
+}
+
+/// For a given set of `Column`s required for predicate `Expr` determine whether all
+/// columns are sorted. Sorted columns may be queried more efficiently in the presence of
+/// a PageIndex.
+fn columns_sorted(
+    _columns: &HashSet<Column>,
+    _schema: &Schema,
+    _metadata: &ParquetMetaData,
+) -> Result<bool> {
+    // TODO How do we know this?
```

Comment:
If the file with `page-index`, it will stored at `ParquetMetaData .page_indexes.boundary_order`.
I think if without page-index there is no way to find out🤔

Reply:
Parquet doesn't have a way to express multi-column sort orders that I am aware of. I thought DataFusion had some plumbing for this, but I think it doesn't extend into the physical scan operators.